src/ports/linux/mib/system_mib.c

```diff
@@ -0,0 +1,575 @@
+         sysoid,
+         OID_LENGTH (sysoid) * sizeof (oid));
+      break;
+
+   default:
+      /* we should never get here, so this is a really bad error */
+      LOG_ERROR (
+         PF_SNMP_LOG,
+         "system_mib(%d): unknown mode (%d)\n",
+         __LINE__,
+         reqinfo->mode);
+      return SNMP_ERR_GENERR;
+   }
+
+   return SNMP_ERR_NOERROR;
+}
+
+int handle_sysUpTime (
+   netsnmp_mib_handler * handler,
+   netsnmp_handler_registration * reginfo,
+   netsnmp_agent_request_info * reqinfo,
+   netsnmp_request_info * requests)
+{
+   /* We are never called for a GETNEXT if it's registered as a
+      "instance", as it's "magically" handled for us.  */
+
+   /* a instance handler also only hands us one request at a time, so
+      we don't need to loop over a list of requests; we'll only get one. */
+
+   switch (reqinfo->mode)
+   {
+
+   case MODE_GET:
+      snmp_set_var_typed_integer (
+         requests->requestvb,
+         ASN_TIMETICKS,
+         netsnmp_get_agent_uptime());
+      break;
+
+   default:
+      /* we should never get here, so this is a really bad error */
+      LOG_ERROR (
+         PF_SNMP_LOG,
+         "system_mib(%d): unknown mode (%d)\n",
+         __LINE__,
+         reqinfo->mode);
+      return SNMP_ERR_GENERR;
+   }
+
+   return SNMP_ERR_NOERROR;
+}
+
+int handle_sysContact (
+   netsnmp_mib_handler * handler,
+   netsnmp_handler_registration * reginfo,
+   netsnmp_agent_request_info * reqinfo,
+   netsnmp_request_info * requests)
+{
+   int ret;
+   pnet_t * pnet = reginfo->my_reg_void;
+   pf_snmp_system_contact_t syscontact;
+   static pf_snmp_system_contact_t * old_syscontact = NULL;
+
+   /* We are never called for a GETNEXT if it's registered as a
+      "instance", as it's "magically" handled for us.  */
+
+   /* a instance handler also only hands us one request at a time, so
+      we don't need to loop over a list of requests; we'll only get one. */
+
+   switch (reqinfo->mode)
+   {
+
+   case MODE_GET:
+      pf_snmp_get_system_contact (pnet, &syscontact);
+      snmp_set_var_typed_value (
+         requests->requestvb,
+         ASN_OCTET_STR,
+         syscontact.string,
+         strlen (syscontact.string));
+      break;
+
+   /*
+    * SET REQUEST
+    *
+    * multiple states in the transaction.  See:
+    * http://www.net-snmp.org/tutorial-5/toolkit/mib_module/set-actions.jpg
+    */
+   case MODE_SET_RESERVE1:
+      ret = netsnmp_check_vb_type_and_max_size (
+         requests->requestvb,
+         ASN_OCTET_STR,
+         sizeof (syscontact.string) - 1);
+      if (ret != SNMP_ERR_NOERROR)
+      {
+         netsnmp_set_request_error (reqinfo, requests, ret);
+      }
+      break;
+
+   case MODE_SET_RESERVE2:
+      old_syscontact = SNMP_MALLOC_TYPEDEF (pf_snmp_system_contact_t);
+      if (old_syscontact == NULL)
+      {
+         netsnmp_set_request_error (
+            reqinfo,
+            requests,
+            SNMP_ERR_RESOURCEUNAVAILABLE);
```

Comment:
SNMP_ERR_RESOURCEUNAVAILABLE seems to be an SNMPv2 error. Profinet uses SNMPv1. Will this work for SNMPv1?

Reply:
Yes, it gets converted to SNMP_ERR_GENERR by the net-snmp stack. When forcing the error:

```
$ snmpset -v1 -c public localhost SNMPv2-MIB::sysContact.0 = bar
Error in packet.
Reason: (genError) A general failure occured
Failed object: SNMPv2-MIB::sysContact.0

$ snmpset -v2c -c public localhost SNMPv2-MIB::sysContact.0 = bar
Error in packet.
Reason: resourceUnavailable (This is likely a out-of-memory failure within the agent)
Failed object: SNMPv2-MIB::sysContact.0

```